ui/org.eclipse.pde.genericeditor.extension/src/org/eclipse/pde/internal/genericeditor/target/extension/model/RepositoryCache.java

```diff
@@ -31,41 +42,73 @@
  */
 public class RepositoryCache {
 
-	private static RepositoryCache instance;
-
-	Map<String, List<UnitNode>> cache = new HashMap<>();
-
 	private RepositoryCache() {
 		//avoid instantiation
 	}
 
-	/**
-	 * @return default instance of this cache.
-	 */
-
-	public static RepositoryCache getDefault() {
-		if (instance == null) {
-			instance = new RepositoryCache();
-		}
-		return instance;
-	}
+	private static final Map<String, CompletableFuture<Map<String, List<IVersionedId>>>> CACHE = new ConcurrentHashMap<>();
 
 	/**
 	 * Fetches information and caches it.
-	 *
-	 * @param repo
-	 *            repository URL
-	 * @param flush
-	 *            whether a flush is needed
-	 * @return list of IUs available in the 'repo' repository. Never
-	 *         <code>null</code>.
+	 * <p>
+	 * All available IUs are returned as a map mapping the IDs of all IUs
+	 * available in the {@code repositories} to the list of all available
+	 * {@link IVersionedId versioned IDs} for that ID. All keys are sorted in
+	 * alphabetical order and all versions are sorted in descending order.
+	 * </p>
+	 * 
+	 * @return all available units in the specified {@code repository} in a map
+	 *         mapping all IDs to all available versions.
 	 */
-	public List<UnitNode> fetchP2UnitsFromRepo(String repo, boolean flush) {
-		if ((flush) || (cache.get(repo) == null)) {
-			List<UnitNode> units = P2Fetcher.fetchAvailableUnits(repo);
-			cache.put(repo, units);
+	public static Map<String, List<IVersionedId>> fetchP2UnitsFromRepos(List<String> repositories) {
+		if (repositories.size() == 1) {
+			try {
+				return fetchP2DataOfRepo(repositories.get(0)).get();
+			} catch (InterruptedException | ExecutionException e) {
+				return Map.of();
+			}
 		}
-		return cache.get(repo);
+		List<Future<Map<String, List<IVersionedId>>>> repos = repositories.stream()
+				.map(RepositoryCache::fetchP2DataOfRepo).toList();
+		// Fetch all repos at once to await pending metadata in parallel
+		return toSortedMap(repos.stream().<Map<String, List<IVersionedId>>>map(f -> {
+			try {
+				return f.get();
+			} catch (InterruptedException | ExecutionException e) {
+				return Map.of();
+			}
+		}).map(Map::values).flatMap(Collection::stream).flatMap(List::stream));
+	}
+
+	public static void prefetchP2MetadataOfRepository(String repository) {
+		fetchP2DataOfRepo(repository);
+	}
+
+	private static Future<Map<String, List<IVersionedId>>> fetchP2DataOfRepo(String repository) {
+		return CACHE.compute(repository, (r, f) -> {
+			if (f != null && f.isDone() && !f.isCompletedExceptionally() && !f.isCancelled()) {
```

Comment:
Job cancellation seem to to result in future cancellation ....

By the way this seems a good use-case for
- https://github.com/eclipse-platform/eclipse.platform/pull/327

Reply:
> By the way this seems a good use-case for
> 
>     * [Support executing tasks with the Job Framework using CompletableFuture eclipse-platform/eclipse.platform#327](https://github.com/eclipse-platform/eclipse.platform/pull/327)

That would help here, yes.
> 
> 
> Beside that the condition seems to be wrong here. This will result in a new future be created for each call if the previous one is not yet finished. Instead of having one big if, I would separate the cases:
> 
> ```
> if (f != null) {
>    if (!f.isDone()) {
>      //not finished yet...
>      return f;
>    }
>    if(!f.isCompletedExceptionally() && !f.isCancelled()) {
>      //finished without an error
>     return f;
>    }
> }
> ```

Indeed it was wrong, thanks for pointing that out.
But I'm not sure if necessary to break it up so much. I added a combined comment that should help if it's not clear.